Add TabRepo2 Models

Innixma opened the pull request:
*Issue #, if available:*

*Description of changes:*

Add TabRepo 2.0 model running logic + model code

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

geoalgo commented on examples/run_quickstart_new_config.py:
```diff
@@ -0,0 +1,156 @@
+            num_bag_folds=8,  # num_bag_folds=8 was used in the TabRepo 2024 paper
+            time_limit=3600,  # time_limit=3600 was used in the TabRepo 2024 paper
+        ),
+        AGModelBagExperiment(
+            # Reproduces the `XGBoost_c1_BAG_L1` model from the TabRepo 2024 paper
+            # The results may slightly differ since we are using a different XGBoost version.
+            name="XGBoost_c1_BAG_L1_Reproduced",
+            model_cls=XGBoostModel,
+            model_hyperparameters={},
+            num_bag_folds=8,
+            time_limit=3600,
+        ),
```
Unwanted duplicate?

Innixma replied:
fair, removed

geoalgo commented on examples/run_quickstart_new_config.py:
```diff
@@ -0,0 +1,156 @@
+        ),
+
+        # This will be a `baseline` in EvaluationRepository, because it doesn't compute out-of-fold predictions and thus can't be used for post-hoc ensemble.
+        Experiment(  # Generic wrapper for any model, including non-AutoGluon models
```
Could we make the naming more precise here and similar to the one above? Maybe `AGModelSingleExperiment`?

Innixma replied:
Experiment is fully generic, and is not tied to AG in any way, so I prefer to keep the name separated from AG.

I already have `AGModelExperiment` for generic AG runs. `Experiment` is the most generic as it could be anything (for example, H2OAutoML or native TabPFNv2)

geoalgo commented on examples/run_quickstart_new_config.py:
```diff
@@ -0,0 +1,156 @@
+    with pd.option_context("display.max_rows", None, "display.max_columns", None, "display.width", 1000):
+        print(f"Config Metrics Example:\n{metrics.head(100)}")
+
+    # Requires `autogluon_bench` and `autogluon_benchmark`
+    # FIXME: AG-Bench is saving unnecessary files in `data/results/output/openml/None/...`
+    # FIXME: AG-Bench might not work out of the box due to the existence of `run_evaluation_openml.evaluate_amlb_results`
+    evaluation_save_dir = os.path.join(expname, "evaluation")
+    evaluator_output = evaluator.plot_overall_rank_comparison(
+        results_df=metrics,
+        save_dir=evaluation_save_dir,
+        evaluator_kwargs={"treat_folds_as_datasets": True},
+    )
```
Could we try import here? It seems like we dont want to fail the run if this table is not plotted since we still have one above and the dependency is optional?

Innixma replied:
I'm leaving this for a future PR, as I eventually want to remove the dependency on AG-Bench and AG-Benchmark by moving the required logic into TabRepo directly. TabRepo should ideally replace AG-Benchmark, as AG-Benchmark was my old code for running experiments prior to making TabRepo 2.0.

geoalgo commented on tabrepo/benchmark/models/wrapper/abstract_class.py:
```diff
@@ -0,0 +1,122 @@
+from autogluon.core.data.label_cleaner import LabelCleaner, LabelCleanerDummy
+from autogluon.core.metrics import get_metric, Scorer
+from autogluon.features import AutoMLPipelineFeatureGenerator
+from autogluon_benchmark.utils.time_utils import Timer
```
Can we avoid this? This would make it a strong requirements? (We have a timer class in TabRepo right?)

Innixma replied:
Agreed, but for now I am keeping. As mentioned above, I will soon remove the dependency on AG-Bench and AG-Benchmark entirely.

Innixma replied:
This will be its own dedicated PR

geoalgo reviewed (commented):
Really cool to have this, looks good to me. I left some comments but I did not get a chance to read carefully all the code yet.

geoalgo commented on examples/run_quickstart_new_config.py:
```diff
@@ -0,0 +1,156 @@
+if __name__ == '__main__':
+    # Load Context
+    context_name = "D244_F3_C1530_10"  # 10 smallest datasets. To run larger, set to "D244_F3_C1530_200"
+    expname = "./experiments/quickstart_new_config"  # folder location to save all experiment artifacts
```
Would be nice to avoid a relative location:
```suggestion
    expname = "Path(__file__).parent / "experiments" / "quickstart_new_config"  # folder location to save all experiment artifacts
```

Innixma replied:
There was a typo in the code suggestion with extra `"`, but thanks for mentioning. I updated the code to address.

geoalgo commented on tabrepo/benchmark/experiment/experiment_constructor.py:
```diff
@@ -0,0 +1,194 @@
+
+class Experiment:
+    """
+
```
Could you add a docstring with the context you give in the example?

Innixma replied:
added docstrings to Experiment + subclasses

Innixma commented:
Thanks for the review @geoalgo! I addressed the comments. I will merge the PR to avoid too much code backup, but if you have additional comments please message me via mattermost so I am aware and i can make a follow-up PR to address additional comments.
